glommio/src/io/read_result.rs

```diff
@@ -2,104 +2,139 @@
 // mit/apache-2.0 license, at your convenience
 //
 // this product includes software developed at datadog (https://www.datadoghq.com/). copyright 2020 datadog, inc.
-//
 use crate::sys::DmaBuffer;
-use std::io;
+use core::num::NonZeroUsize;
 use std::rc::Rc;
 
-#[derive(Debug, Clone)]
-/// ReadResult encapsulates a buffer, returned by read operations like [`get_buffer_aligned`] and
-/// [`read_at`]
-///
-/// [`get_buffer_aligned`]: struct.DmaStreamReader.html#method.get_buffer_aligned
-/// [`read_at`]: struct.DmaFile.html#method.read_at
+#[derive(Clone, Debug)]
+/// ReadResult encapsulates a buffer, returned by read operations like
+/// [`get_buffer_aligned`](super::DmaStreamReader::get_buffer_aligned) and
+/// [`read_at`](super::DmaFile::read_at)
 pub struct ReadResult {
-    buffer: Option<Rc<DmaBuffer>>,
+    inner: Option<ReadResultInner>,
+}
+
+impl core::ops::Deref for ReadResult {
+    type Target = [u8];
+
+    /// Allows accessing the contents of this buffer as a byte slice
+    fn deref(&self) -> &[u8] {
+        self.inner.as_ref().map(|inner| &**inner).unwrap_or(&[])
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ReadResultInner {
+    buffer: Rc<DmaBuffer>,
     offset: usize,
-    end: usize,
+    len: NonZeroUsize,
```

Comment:
This (usage of `NonZeroUsize`) is probably needed to make sure that `rustc` doesn't need to reserve additional memory for the surrounding `Option` enum tag.

Reply:
Are we sure that the `NonZeroUsize` in the `len` field is actually gonna help optimize the layout? The example given in the docs refer to cases where it's wrapped in an option directly but not a case like this.. So do you have any documentation or even measured example of this working to optimize the layout and remove the need for the extra space for the tag?

Reply:
I added an unit test for this, thanks.

Reply:
Shouldn't those things go into comments in the code?

For github, once the PR is closed the information is lost